File: go/0875.koko-eating-bananas/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:34
// leetgo: dev
// https://leetcode.com/problems/koko-eating-bananas/

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func minEatingSpeed(piles []int, h int) int {
	min := int(math.Ceil(float64(sliceSum(piles)) / float64(h))) // left (smallest)
	max := slices.Max(piles)                                     // right

	best := max

	for min <= max {
		k := int((min + max) / 2)
		hours := 0

		for _, pile := range piles {
			hours += int(math.Ceil(float64(pile) / float64(k)))
		}

		if hours <= h {
			best = int(math.Min(float64(best), float64(k)))
			max = k - 1
		} else {
			min = k + 1
		}
	}

	return best
}

func sliceSum(slice []int) int {
	sum := 0
	for _, val := range slice {
		sum += val
	}
	return sum
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	piles := Deserialize[[]int](ReadLine(stdin))
	h := Deserialize[int](ReadLine(stdin))
	ans := minEatingSpeed(piles, h)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0875.koko-eating-bananas/solution.rs
// Created by Olgierd Palasz at 2026/08/09 20:36
// leetgo: dev
// https://leetcode.com/problems/koko-eating-bananas/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin

impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut start = 1;
        let mut end = *piles.iter().max().unwrap();

        while start < end {
            let mid = start + (end - start) / 2;

            let amount: i32 = piles.iter().map(|&x| (x + mid - 1) / mid).sum();

            if amount > h {
                start = mid + 1;
            } else {
                end = mid;
            }
        }

        start
    }
}

// @lc code=end
fn main() -> Result<()> {
    let piles: Vec<i32> = deserialize(&read_line()?)?;
    let h: i32 = deserialize(&read_line()?)?;
    let ans: i32 = Solution::min_eating_speed(piles, h).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
